torch/_dynamo/optimizations/training.py

```diff
@@ -277,18 +277,62 @@ def prims_executor(gm, inputs, *, executor):
     return partial(execute, prim_gm, executor=executor)
 
 
+def nvprims_fw_bw_partition_fn(joint_module, joint_inputs):
+    # This function is called once per forward+backward pass of a graph in AOT
+    # Autograd. We use it to set up the nvFuser-specific FX graph that is later
+    # passed to the executor.
+    from functorch.compile import min_cut_rematerialization_partition
+
+    from torch._prims.context import TorchRefsNvfuserCapabilityMode
+    from torch.fx.experimental.proxy_tensor import make_fx
+
+    # AOT Autograd expects arguments of the traced function to be named exactly
+    # "primals, tangents"
+    def func(primals, tangents):
+        return joint_module(primals, tangents)
+
+    # First we trace the graph conditionally decomposing nodes
+    # that can be sent to the nvfuser executor
+    with TorchRefsNvfuserCapabilityMode():
```

Comment:
Does this mean that later we are speculatively lowering the partitioned graph to nvprim again in `prims_executor`?

should we just skip the speculative lowering, or is the second lowering there to catch some other decomposed op?

Reply:
Unfortunately yes.

Initially, I did leave this only in the partitioner code, but unfortunately, AOT has two different code paths for the case when at least one of the inputs requires grad and none requires grad, for the latter case "partition_fn" is not used at all, and there's no way to pass that information further without modifying AOT code. All inputs to the "fw_compiler" functions are detached and do not require grad, so it's not possible to determine this from within there.

Reply:
Okay with a bit of monkey patching unnecessary lowering step is avoided in https://github.com/pytorch/pytorch/pull/88204/commits/2a1103c918db8a451dbfe5e4c2981358fa17ed73

Reply:
Looks good~ thx for patching this~

Reply:
A quick follow up question.

So now we are not lowering to nvprims after aot_autograd, does this mean `post-autograd` decomposition won't be lowered to nvprims now :exploding_head:

Reply:
I don't know exactly when `post-autograd` decomposition happens.. Hopefully it's not the case and we can still use it.
Can we get a CI test to guard/verify that behavior?

Reply:
We don't use decompositions (`aot_module_simplified(decompositions=None)`). Even if we used it, it would happen before calling the partitioning function.